stan/math/rev/functor/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,248 @@
+#ifndef STAN_MATH_REV_FUNCTOR_APPLY_SCALAR_BINARY_HPP
+#define STAN_MATH_REV_FUNCTOR_APPLY_SCALAR_BINARY_HPP
+
+#include <stan/math/prim/fun/as_column_vector_or_scalar.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/prim/err/check_matching_dims.hpp>
+#include <stan/math/prim/err/check_matching_sizes.hpp>
+#include <stan/math/prim/fun/num_elements.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Specialisation for use with combinations of
+ * `Eigen::Matrix` and `var_value<Eigen::Matrix>` inputs.
+ * Eigen's binaryExpr framework is used for more efficient indexing of both row-
+ * and column-major inputs  without separate loops.
+ *
+ * @tparam T1 Type of first argument to which functor is applied.
+ * @tparam T2 Type of second argument to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x First Matrix input to which operation is applied.
+ * @param y Second Matrix input to which operation is applied.
+ * @param f functor to apply to Matrix inputs.
+ * @return `var_value<Matrix>` with result of applying functor to inputs.
+ */
+template <typename T1, typename T2, typename F,
+          require_any_var_matrix_t<T1, T2>* = nullptr,
+          require_all_matrix_t<T1, T2>* = nullptr>
```

Comment:
Do these all need to be separate overloads? Given that the only difference between the non-nested calls is the use of `check_matching_dims` and `check_matching_sizes`, could you just use a handful of overloads with a more liberal use of `require_any_`?

Reply:
Ah yes! So I made a new meta function `is_container_or_var_matrix` that lets us re-use the stuff already in `prim`.